client/landing/stepper/declarative-flow/internals/hooks/use-step-navigation-with-tracking/index.ts

```diff
@@ -12,33 +14,35 @@ interface Params< FlowSteps extends StepperStep[] > {
 }
 
 export const useStepNavigationWithTracking = ( {
-	flow,
+	flow: { useTracksEventProps, useStepNavigation, name: flowName, variantSlug },
 	currentStepRoute,
 	navigate,
 	steps,
 }: Params< ReturnType< Flow[ 'useSteps' ] > > ) => {
-	const stepNavigation = flow.useStepNavigation(
+	const stepNavigation = useStepNavigation(
 		currentStepRoute,
 		navigate,
 		steps.map( ( step ) => step.slug )
 	);
 	const intent =
 		useSelect( ( select ) => ( select( ONBOARD_STORE ) as OnboardSelect ).getIntent(), [] ) ?? '';
+	const tracksEventPropsFromFlow = useTracksEventProps?.();
 
-	const shouldTrackSubmit = flow.use__Temporary__ShouldTrackEvent?.( 'submit' ) ?? false;
-
-	return {
-		...stepNavigation,
-		submit: ( providedDependencies: ProvidedDependencies = {}, ...params: string[] ) => {
-			shouldTrackSubmit &&
+	return useMemo(
+		() => ( {
+			...stepNavigation,
+			submit: ( providedDependencies: ProvidedDependencies = {}, ...params: string[] ) => {
 				recordSubmitStep(
 					providedDependencies,
 					intent,
-					flow.name,
+					flowName,
 					currentStepRoute,
-					flow.variantSlug
+					variantSlug,
+					tracksEventPropsFromFlow?.submit?.[ STEPPER_TRACKS_EVENT_STEP_NAV_SUBMIT ]
```

Comment:
Commenting here for perspective over the new code. The error should be addressed now. 

You'll notice a couple of changes. I switched the approach from "intercepting" single events to passing a more structured single object back with the controls as keys (that we can extend going forward). This feels more natural to me also for adapting to variable events down the line. The alternative, to replace a Tracks event with another falls a little on the ambiguous side to me.

It's also more efficient for applying optimizations over this structure (vs on each individual event object). The only other path is to not memoize anything (neither here nor in the flow end) as it would make the entirety of it ineffective.

Reply:
Great design. Thanks!

Reply:
Good point. It does appear restrictive in that sense. 

My motivation has been to not override things, as that's also restrictive (one may need more props for a default Stepper event and additional events logged whenever those get logged - and that through a clear API).

Would perhaps wrapping these events like we do with "submit", "exitFlow", etc. make sense? It could be anything, like "pageView" for this case. Do we have more default/framework-level Tracks events that don't map to Stepper controls/events?

Reply:
I'll give it more thought too. It is the idea to control and extend everything that the framework tracks, nothing different there. Just thought these canonical groups/objects make it clear (but there's obviously the naming to get right)

Reply:
Ok. I've given this some thought, and, all things considered, I don't quite see an alternative being better right now.

The naming is inevitable, whether it's made explicit in the type or happens on the consuming end e.g.

```
const propsSubmit = useTracksEventProps( "submit event" );
const propsPageView = useTracksEventProps( "page view event" );
... etc.
```

I don't think there's another way to do this while allowing optimizations to work / propagate through to FlowRenderer. 

I suggest we proceed here and keep reworking the type. It should be easy to work this out as we go - see the final form, backtrack if we need to, etc. There's a lot to build around this.

Reply:
It might even make sense to nest these and be super-changed about it e.g.

```
tracksEventPropsFromFlow?.stepNav?.submit
tracksEventPropsFromFlow?.stepNav?.goBack
tracksEventPropsFromFlow?.flow?.start
tracksEventPropsFromFlow?.step?.start
```

Reply:
I'm happy to proceed. 

Just to share the way I'm thinking about this. I see an extension of Tracks that is barely connected to Stepper. I really don't want to specify every type of Stepper event on top of Tracks because then we'll have to maintain that and restrict users within it. 

So I'd vote against nesting for sure. I am not sure if it's a suitable solution, but what I see a simple middleware that intercepts tracks events and augments them. Without being aware of Stepper login like submit and stepNav and such.

Reply:
Cool. I honestly don't mind going back to being a simple extension system of the default events. The idea of a middleware overriding the event is what mainly appears ambiguous and restrictive to me (if we want to go down that path, then the current form is definitely the better alternative, at least for me). I want to have a clear communication line between API and consumer, and replacing events doesn't match.

The only other part about intercepting/passing single events is the bit of cryptic processing that it hinders (like useMemo calls stacking up in the order of events being handled, or whatever else stems from calling a hook N times outside of life cycle).

Reply:
> going back to being a simple extension system of the default events

Let's do that perhaps. Keep it simple for now :-)

Reply:
☝️ in https://github.com/Automattic/wp-calypso/pull/93273/commits/85fbc60c85934d5c08d3c85a34dfbea99db0fce9